File: src/main/java/fi/metatavu/oioi/cm/rest/translate/ResourceTranslator.java
package fi.metatavu.oioi.cm.rest.translate;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import fi.metatavu.oioi.cm.model.KeyValueProperty;
import fi.metatavu.oioi.cm.persistence.model.Resource;
import fi.metatavu.oioi.cm.resources.ResourceController;

/**
 * Translator for Resource REST entity 
 * 
 * @author Antti Leppä
 */
public class ResourceTranslator extends AbstractTranslator<fi.metatavu.oioi.cm.persistence.model.Resource, fi.metatavu.oioi.cm.model.Resource> {
  
  @Inject
  private ResourceController resourceController;

  @Override
  public fi.metatavu.oioi.cm.model.Resource translate(fi.metatavu.oioi.cm.persistence.model.Resource entity) {
    fi.metatavu.oioi.cm.model.Resource result = new fi.metatavu.oioi.cm.model.Resource();    
    result.setId(entity.getId());
    result.setData(entity.getData());
    result.setName(entity.getName());
    result.setParentId(entity.getParent() != null ? entity.getParent().getId() : null);
    result.setSlug(entity.getSlug());
    result.setType(entity.getType());
    result.setProperties(getProperties(entity));
    result.setStyles(getStyles(entity));
    result.setCreatedAt(entity.getCreatedAt());
    result.setCreatorId(entity.getCreatorId());
    result.setLastModifierId(entity.getLastModifierId());
    result.setModifiedAt(entity.getModifiedAt());
    return result;
  }

  /**
   * Translates styles to REST format
   * 
   * @param entity resource
   * @return styles as REST key value pairs
   */
  private List<KeyValueProperty> getStyles(Resource entity) {
    return resourceController.listStyles(entity).stream().map(resourceStyle -> {
      KeyValueProperty result = new KeyValueProperty();
      result.setKey(resourceStyle.getKey());
      result.setValue(resourceStyle.getValue());
      return result;
    }).collect(Collectors.toList());
  }

  /**
   * Translates properties to REST format
   * 
   * @param entity resource
   * @return properties as REST key value pairs
   */
  private List<KeyValueProperty> getProperties(Resource entity) {
    return resourceController.listProperties(entity).stream().map(resourceProperty -> {
      KeyValueProperty result = new KeyValueProperty();
      result.setKey(resourceProperty.getKey());
      result.setValue(resourceProperty.getValue());
      return result;
    }).collect(Collectors.toList());
  }
  
}

File: src/main/kotlin/fi/metatavu/oioi/cm/rest/translate/ResourceTranslator.kt
package fi.metatavu.oioi.cm.rest.translate

import fi.metatavu.oioi.cm.model.KeyValueProperty
import fi.metatavu.oioi.cm.persistence.model.Resource
import fi.metatavu.oioi.cm.persistence.model.ResourceProperty
import fi.metatavu.oioi.cm.persistence.model.ResourceStyle
import fi.metatavu.oioi.cm.resources.ResourceController
import javax.enterprise.context.ApplicationScoped
import javax.inject.Inject

/**
 * Translator for Resource REST entity
 *
 * @author Jari Nykänen
 * @author Antti Leppä
 */
@ApplicationScoped
class ResourceTranslator: AbstractTranslator<Resource, fi.metatavu.oioi.cm.model.Resource>() {

    @Inject
    private lateinit var resourceController: ResourceController

    override fun translate(entity: Resource): fi.metatavu.oioi.cm.model.Resource {
        val resourceStyles = resourceController.listStyles(entity)
        val resourceProperties = resourceController.listProperties(entity)

        val modifiedAt = resourceStyles.mapNotNull { it.modifiedAt }
            .plus(resourceProperties.mapNotNull{ it.modifiedAt })
            .plus(entity.modifiedAt!!)
            .maxOf { it }

        return fi.metatavu.oioi.cm.model.Resource(
            id = entity.id,
            data = entity.data,
            name = entity.name!!,
            orderNumber = entity.orderNumber,
            parentId = entity.parent?.id,
            slug = entity.slug!!,
            type = entity.type!!,
            properties = getProperties(resourceProperties),
            styles = getStyles(resourceStyles),
            createdAt = entity.createdAt,
            creatorId = entity.creatorId,
            lastModifierId = entity.lastModifierId,
            modifiedAt = modifiedAt
        )
    }

    /**
     * Translates styles to REST format
     *
     * @param resourceStyles resource styles
     * @return styles as REST key value pairs
     */
    private fun getStyles(resourceStyles: List<ResourceStyle>): List<KeyValueProperty> {
        return resourceStyles.map { resourceStyle: ResourceStyle ->
            KeyValueProperty(
                key = resourceStyle.key!!,
                value = resourceStyle.value!!
            )
        }
    }

    /**
     * Translates properties to REST format
     *
     * @param resourceProperties resource properties
     * @return properties as REST key value pairs
     */
    private fun getProperties(resourceProperties: List<ResourceProperty>): List<KeyValueProperty> {
        return resourceProperties.map { resourceProperty: ResourceProperty ->
            KeyValueProperty(
                key = resourceProperty.key!!,
                value = resourceProperty.value!!
            )
        }
    }

}
